src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+import asyncio
+import sys
+import time
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar, Union, cast
+
+if sys.version_info < (3, 10):
+    from typing_extensions import ParamSpec
+else:
+    from typing import ParamSpec
+
+Numeric = Union[int, float]
+P = ParamSpec("P")
+T = TypeVar("T")
+A = TypeVar("A", bound=Callable[..., Awaitable[Any]])
+
+
+class UnavailableTokensError(Exception):
+    pass
+
+
+class TokenLimiter:
+    """
+    A simple in-memory rate-limiter implemented using the leaky bucket algorithm.
+    """
+
+    def __init__(
+        self,
+        per_minute_rate: Numeric,
+        starting_tokens: Numeric,
+        max_tokens: Numeric,
+        rate_multiplier: Numeric = 1,
+    ):
+        self.rate_multiplier = rate_multiplier
+        self.rate = self._per_second_rate(per_minute_rate)
+        self.tokens = starting_tokens
+        self.max_tokens = max_tokens
+        self.created = time.time()
+        self.last_checked = self.created
+        self.total_tokens: Numeric = 0
+
+    def _per_second_rate(self, per_minute_rate: Numeric) -> float:
+        return round(per_minute_rate / 60, 3) * self.rate_multiplier
+
+    def refresh_limit(self, per_minute_rate: Numeric, max_tokens: Numeric) -> None:
+        new_rate = self._per_second_rate(per_minute_rate)
+        if self.rate != new_rate:
+            self.rate = new_rate
+            self.tokens = 0  # reset tokens as conservatively as possible
+            self.max_tokens = max_tokens
+            self.last_checked = time.time()
+
+    def available_tokens(self) -> float:
+        time_since_last_checked = time.time() - self.last_checked
+        return min(self.max_tokens, self.rate * time_since_last_checked + self.tokens)
+
+    def spend_tokens_if_available(self, token_cost: Numeric) -> None:
+        if token_cost > self.available_tokens():
+            raise UnavailableTokensError
+        now = time.time()
+        current_tokens = min(self.max_tokens, self.tokens + (now - self.last_checked) * self.rate)
+        self.tokens = current_tokens - token_cost
+        self.last_checked = now
+        self.total_tokens += token_cost
+
+    def spend_tokens(self, token_cost: Numeric) -> None:
+        self.tokens -= token_cost
+        self.total_tokens += token_cost
+
+    def wait_for_then_spend_available_tokens(self, token_cost: Numeric) -> None:
+        MAX_WAIT_TIME = 5 * 60  # 5 minutes
```

Comment:
Kinda confused why this is not a constant outside of the function scope or not just an optional parameter - seems like a magic number?

Reply:
It's just a magic number to prevent an infinite loop, yeah, I'll make it an optional parameter